app/src/main/java/com/example/mystore/ui/home/HomeViewModel.kt

```diff
@@ -16,33 +20,106 @@ import javax.inject.Inject
 class HomeViewModel @Inject constructor(
     private val apiRepository: IApiRepository
 ) : ViewModel() {
+    private var allProducts = mutableStateListOf<ProductModel>()
     var products = mutableStateListOf<ProductModel>()
     var isLoading by mutableStateOf(false)
+    var categories = mutableStateListOf<CategoryDetailsModel>()
+
+    var input by mutableStateOf("")
+    private var priceStartRange by mutableFloatStateOf(20f)
+    private var priceEndRange by mutableFloatStateOf(150f)
+    private var rating by mutableIntStateOf(4)
+    private var selectedCategories = mutableStateListOf<String>().apply {
+        addAll(listOf("Home"))
+    }
+    var filtersCount by mutableIntStateOf(3)
 
     init {
         getProducts()
+        getCategories()
+    }
+
+    fun filterProducts(newInput: String? = null) {
+        products.apply {
+            clear()
+            addAll(allProducts.filter {
+                it.price in (priceStartRange..priceEndRange) && it.rating <= rating
+            })
+        }
+
+        if (selectedCategories.isNotEmpty() && selectedCategories.count() != categories.count()) {
+            products.apply {
```

Comment:
No need for `.apply` here as you're invoking a single function. A simple `products.removeAll` will result in the exact same thing with fewer code.